presto-hive/src/main/java/com/facebook/presto/hive/HiveMetadata.java

```diff
@@ -334,6 +339,82 @@ private ConnectorTableMetadata getTableMetadata(SchemaTableName tableName)
         return columns.build();
     }
 
+    @Override
+    public TableStatistics getTableStatistics(ConnectorSession session, ConnectorTableHandle tableHandle, Constraint<ColumnHandle> constraint)
+    {
+        List<ConnectorTableLayoutResult> tableLayouts = getTableLayouts(session, tableHandle, constraint, Optional.empty());
+        if (tableLayouts.isEmpty()) {
+            // todo - return statistics matching empty table instead
+            return EMPTY_STATISTICS;
+        }
+        HiveTableLayoutHandle layout = checkType(tableLayouts.get(0), HiveTableLayoutHandle.class, "layoutHandle");
+        List<HivePartition> hivePartitions = layout.getPartitions().orElse(ImmutableList.of());
+        List<Long> knownPartitionRowCounts = hivePartitions.stream()
+                .map(this::getPartitionStatistics)
+                .map(PartitionStatistics::getNumRows)
+                .filter(Optional::isPresent)
+                .map(Optional::get)
+                .collect(toList());
+
+        long knownPartitionRowCountsSum = knownPartitionRowCounts.stream().mapToLong(a -> a).sum();
+        long partitionsWithStatsCount = knownPartitionRowCounts.size();
+        long allPartitionsCount = hivePartitions.size();
+
+        TableStatistics tableStatistics = EMPTY_STATISTICS;
+        if (partitionsWithStatsCount > 0) {
+            tableStatistics = TableStatistics.builder()
+                    .setRowCount(new Estimate(1.0 * knownPartitionRowCountsSum / partitionsWithStatsCount * allPartitionsCount))
+                    .build();
+        }
+        return tableStatistics;
+    }
+
+    private PartitionStatistics getPartitionStatistics(HivePartition hivePartition)
+    {
+        String databaseName = hivePartition.getTableName().getSchemaName();
+        String tableName = hivePartition.getTableName().getTableName();
+        String partitionId = hivePartition.getPartitionId();
+        if (partitionId.equals(HivePartition.UNPARTITIONED_ID)) {
+            Optional<Table> tableOptional = metastore.getTable(databaseName, tableName);
+            Table table = tableOptional
+                    .orElseThrow(() -> new IllegalArgumentException(format("Could not get metadata for table %s.%s", databaseName, tableName)));
+            return readStatisticsFromParameters(table.getParameters());
+        }
+        else {
+            Optional<Partition> partitionOptional = metastore.getPartition(databaseName, tableName, toPartitionValues(partitionId));
```

Comment:
> Also, we should probably encapsulate the logic for getting stats so that it can be overridden in connectors that extend this one. For example, at FB, we'd need to call a different service that maintains those stats that is separate from the metastore.

So you suggest to move statistics gathering implementation out of HiveMetadata class.
And make `HiveMetadata.getTableStatistics` just a delegate to some method in `HiveStatisticsProvider`.
Default implementation of `HiveStatisticsProvider` would do the logic currently present in `HiveMetadata` and you at FB could inject different implementation which matches your infra. 

Should not be too hard. Would it be fine to implement this as refactoring commit on top of commit currently in PR?
It will be easier to not rebase everything on top of this change.